src/usbhost/device-usbhost.cpp

```diff
@@ -0,0 +1,93 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2015 Intel Corporation. All Rights Reserved.
+
+#ifdef RS2_USE_ANDROID_BACKEND
+
+#include "device-usbhost.h"
+#include "endpoint-usbhost.h"
+#include "interface-usbhost.h"
+#include "usbhost.h"
+#include "../types.h"
+
+#include <string>
+#include <regex>
+#include <sstream>
+#include <mutex>
+
+namespace librealsense
+{
+    namespace platform
+    {
+        usb_device_usbhost::usb_device_usbhost(::usb_device* handle) :
+            _handle(handle)
+        {
+            usb_descriptor_iter it;
+            ::usb_descriptor_iter_init(handle, &it);
+            usb_descriptor_header *h = usb_descriptor_iter_next(&it);
+            librealsense::platform::usb_spec conn_spec;
+            if (h != nullptr && h->bDescriptorType == USB_DT_DEVICE) {
+                usb_device_descriptor *device_descriptor = (usb_device_descriptor *) h;
+                conn_spec = librealsense::platform::usb_spec(device_descriptor->bcdUSB);
+            }
+
+            do {
+                h = usb_descriptor_iter_next(&it);
+                if(h == NULL)
+                    break;
+                if (h->bDescriptorType == USB_DT_INTERFACE_ASSOCIATION) {
+                    auto iad = *(usb_interface_assoc_descriptor *) h;
+                }
+                if (h->bDescriptorType == USB_DT_INTERFACE) {
+                    auto id = *(usb_interface_descriptor *) h;
+                    auto in = id.bInterfaceNumber;
+                    _interfaces[in] = std::make_shared<usb_interface_usbhost>(id, it);
+                    if(id.bInterfaceSubClass != USB_SUBCLASS_CONTROL && id.bInterfaceSubClass != USB_SUBCLASS_HWM)
+                        continue;
+                    usb_device_info info{};
+                    info.vid = usb_device_get_vendor_id(_handle);
+                    info.pid = usb_device_get_product_id(_handle);
+                    info.id = std::string(usb_device_get_name(_handle));
+                    info.unique_id = std::string(usb_device_get_name(_handle));
+                    info.mi = in;
+                    info.conn_spec = conn_spec;
+                    _infos.push_back(info);
+                }
+            } while (h != nullptr);
+
+            _usb_device_descriptor = usb_device_get_device_descriptor(_handle);
+        }
+
+        void usb_device_usbhost::release()
+        {
+            _messenger.reset();
```

Comment:
Is this internally mutex-protected to allow flush ?

Reply:
the messenger destructor calls to `_pipe->cancel_request` and joins the pipe thread it should be safe